[service] TP-Link integration

cicoub13 opened the pull request:
### Pull Request check-list

To ensure your Pull Request can be accepted as fast as possible, make sure to review and check all of these items:

- [x] If your changes affects code, did your write the tests? 
- [x] Are tests passing? (`npm test` on both front/server)
- [x] Is the linter passing? (`npm run eslint` on both front/server)
- [x] Did you run prettier? (`npm run prettier` on both front/server)
- [x] If you are adding a new features/services, did you run integration comparator? (`npm run compare-translations` on front)
- [x] If your changes modify the API (REST or Node.js), did you modify the API documentation? (Documentation is based on comments in code)
- [X] If you are adding a new features/services which needs explanation, did you modify the user documentation? See [the GitHub repo](https://github.com/GladysAssistant/v4-website) and the [website](https://gladysassistant.com). https://github.com/GladysAssistant/v4-website/pull/28
- [X] Did you add fake requests data for the demo mode (`front/src/config/demo.json`) so that the demo website is working without a backend? (if needed) See [https://demo.gladysassistant.com](https://demo.gladysassistant.com).

NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open.

### Description of change

I worked on adding TP Link service to control plugs and bulbs. I used https://github.com/plasticrake/tplink-smarthome-api library to control devices.

Documentation and tests are ✅ 

[GIF of integration working](https://recordit.co/wpaPxyxGpu)

### Main concerns

* I don't like the way I developped discovery service : https://github.com/cicoub13/Gladys/blob/tplink/server/services/tp-link/lib/smart-device/smart-device.getDevices.js#L16 => Tested and working. There is no other way with this devices.
* I can't test with bulbs because I don't have => Done with the help of Will9
* In frontend, I grouped Setup and Devices page because TP Link does not require any setup. Devices are discoverable on the network. => I think it's the good way to handle this kind of devices.

cicoub13 committed:
feat(services): Add tp-link service

cicoub13 committed:
fix(tp-link): Fix bulb discovery

- Better management of device type
- Fix some labels

cicoub13 committed:
docs(tp-link): Add mock devices in demo.json

cicoub13 committed:
feat(tp-link): Write some tests and fix

- Add 100% test coverage
- Fix selector / external_id

atrovato commented on front/src/routes/integration/all/tp-link/device-page/Device.jsx:
```diff
@@ -0,0 +1,60 @@
+    try {
+      await this.props.saveDevice(this.props.device, this.props.deviceIndex);
+    } catch (e) {
+      this.setState({ error: RequestStatus.Error });
```
Here you add error indicator on error, but you never clear it on succes.

cicoub13 replied:
I took code from other services 😞 
I'm not sure error is displayed...

Pierre-Gilles replied:
Any update on this ?

cicoub13 replied:
Solved (error displayed and clear if success)

atrovato commented on front/src/routes/integration/all/tp-link/device-page/Device.jsx:
```diff
@@ -0,0 +1,60 @@
+    try {
+      await this.props.deleteDevice(this.props.device, this.props.deviceIndex);
+    } catch (e) {
+      this.setState({ error: RequestStatus.Error });
```
Here you add error indicator on error, but you never clear it on succes.

Pierre-Gilles replied:
Same, any update on this ? Error handling is really important

cicoub13 replied:
Solved (error displayed and clear if success)

atrovato commented on front/src/routes/integration/all/tp-link/device-page/actions.js:
```diff
@@ -0,0 +1,122 @@
+          options.search = state.tpLinkDeviceSearch;
+        }
+        const tpLinkDevices = await state.httpClient.get('/api/v1/service/tp-link/device', options);
+        const tpLinkDevicesMap = new Map();
```
Can we you object instead of Map?

cicoub13 replied:
I think Map is better to look for devices after (line 47) :
`return !state.tpLinkDevicesMap.has(device.external_id);`

Are you afraid about memory ?

Pierre-Gilles replied:
Are you against Map @atrovato ? 😄 Not the first time I see you writing this comment. 

Map is great! It's indeed better to lookup for a specific key.

atrovato commented on server/package.json:
```diff
@@ -90,13 +90,13 @@
     "node-nlp": "^3.0.3",
     "node-schedule": "^1.3.2",
-    "node-webcrypto-ossl": "^1.0.48",
+    "node-webcrypto-ossl": "^1.0.49",
```
Is this required?
As it is at server side, this can have side effect.

cicoub13 replied:
Fixed

atrovato commented on server/package.json:
```diff
@@ -90,13 +90,13 @@
     "sequelize": "^4.42.0",
     "sequelize-cli": "^5.5.1",
-    "sqlite3": "^4.0.6",
+    "sqlite3": "^4.2.0",
```
Is this required?
As it is at server side, this can have side effects.

cicoub13 replied:
Fixed

atrovato commented on server/services/tp-link/api/tp-link.controller.js:
```diff
@@ -0,0 +1,52 @@
+   * @apiGroup TPLink
+   */
+  async function turnOff(req, res) {
+    await tpLinkSmartDeviceHandler.setValue(req.params.device_id, 0);
```
`turnOn` and `turnOff` looks prettier similar.
For more, I might be wrong, but I can't find any usage.

cicoub13 replied:
You're right. I'll remove them.

Pierre-Gilles replied:
Agree, we don't want that specific services have their own separate REST API. The Gladys native core API is doing exactly the same :) Please remove them!

cicoub13 replied:
Fixed

atrovato commented on server/services/tp-link/lib/smart-device/smart-device.getDevices.js:
```diff
@@ -0,0 +1,50 @@
+    });
+  });
+  return new Promise((resolve) => {
+    setTimeout(() => {
```
Do we really need a timer here?

cicoub13 replied:
The tplink-smarthome-api is not compatible with await/async so I think yes.
After launchin the startDiscovery with a discoveryTimeout, I wait for 2000 ms and send back the devices.
If you find another solution, I'll implement it.

atrovato commented on server/services/tp-link/lib/smart-device/smart-device.setValue.js:
```diff
@@ -0,0 +1,42 @@
+  if (!internalDevice) {
+    throw new NotFoundError(`TP_LINK_DEVICE_NOT_FOUND`);
+  }
+  const state = value === 1 ? TP_LINK_ON : TP_LINK_OFF;
```
Not sure to well understand, you check if value === 1 to set TP_LINK_ON which is 1?
So `const state = value;`?

cicoub13 replied:
I used **constant** variables to easily change if one day TP-Link change the ON/OFF values.
Do you think it's useless ?

Pierre-Gilles replied:
Constants are great! Looks good to me.

atrovato commented on server/services/tp-link/package.json:
```diff
@@ -0,0 +1,20 @@
+    "arm64"
+  ],
+  "dependencies": {
+    "bluebird": "^3.7.2",
```
Relative to timer comment, if timer is requested, you can use bluebird Promise.delay().
But I can't see any usage.

cicoub13 replied:
Fixed

atrovato commented on server/test/services/tp-link/smart-device/smart-device.setValue.test.js:
```diff
@@ -0,0 +1,120 @@
+    // EXECUTE
+    await tpLinkService.device.setValue(devices[1], { category: DEVICE_FEATURE_CATEGORIES.LIGHT }, 1);
+
+    // ASSERT
```
Assert?

cicoub13 replied:
Fixed

atrovato commented on server/test/services/tp-link/smart-device/smart-device.setValue.test.js:
```diff
@@ -0,0 +1,120 @@
+    // EXECUTE
+    await tpLinkService.device.setValue(devices[0], { category: DEVICE_FEATURE_CATEGORIES.SWITCH }, 1);
+
+    // ASSERT
```
Assert?

cicoub13 replied:
Fixed

atrovato reviewed (changes requested):
Nice, not as many files to review ;)
Well written, just some comments.

I didn't test it (no device), I didn't try to start it locally.
So wait for tester feedbacks.

cicoub13 committed:
fix(tp-link): Take PR feedbacks

- Missing asserts in tests
- Usage of bluebird Promise
- Don't update dependencies version
- Remove turnOn and turnOff from Controller

Pierre-Gilles reviewed (changes requested):
Thanks for your PR 🙏 After @atrovato changes, it looks really good! 

I tested on my machine but as I don't have TP-Link devices I only saw the scan page. I had a look on the video, and to me it looks good.

Just one feedback, the `server/package-lock.json` file should not be modified by this PR which is strictly about a service :) Could you remove changes on this file?

After that, it's good for me. 

@atrovato good for you too?

atrovato reviewed (dismissed):
+1

Pierre-Gilles commented on front/src/routes/integration/all/tp-link/device-page/FoundDevices.jsx:
```diff
@@ -0,0 +1,93 @@
+const GITHUB_BASE_URL = 'https://github.com/GladysAssistant/Gladys/issues/new';
+
+const createGithubUrl = device => {
+  const title = encodeURIComponent(`Philips Hue: Add device ${device.model}`);
```
This code is related to Philips Hue integration.

cicoub13 replied:
Fixed

Pierre-Gilles reviewed (changes requested):
Thanks for your PR ! :) I'm really looking forward to merge it in Gladys.

There is still a few comments from @atrovato that were not fixed, mainly on the error handling.

Error handling is really important in Gladys 4. We don't want users to be in a "lost" state when something goes wrong, we want them to understand what happened, and have a clear idea of what's wrong.

Don't hesitate if you have any questions on this :)

cicoub13 committed:
fix(tp-link): Manage front errors when save/delete

Pierre-Gilles commented on front/src/routes/integration/all/tp-link/device-page/Device.jsx:
```diff
@@ -0,0 +1,75 @@
+  saveDevice = async () => {
+    this.setState({ loading: true });
+    this.setState({ saveError: null });
+    this.setState({ deleteError: null });
```
Why don't you do all this in one setState?

```javascript
this.setState({
  loading: true,
  saveError: null,
  deleteError: null
});
```

setState is an expensive operation as it triggers a whole front pain. Even if preact has optimization to avoid painting the screen all the time, it's still better to do one setState instead of 3 :)

cicoub13 replied:
Fixed (I'm not used to Frontend development but learning every day 😄)

Pierre-Gilles commented on front/src/routes/integration/all/tp-link/device-page/Device.jsx:
```diff
@@ -0,0 +1,75 @@
+  deleteDevice = async () => {
+    this.setState({ loading: true });
+    this.setState({ saveError: null });
+    this.setState({ deleteError: null });
```
Same here!

cicoub13 replied:
Fixed

Pierre-Gilles reviewed (changes requested):
Thanks for your changes! I still have a very small 10 seconds feedback, and after that I'm good to merge! :) 

Thanks again for this amazing work 👏

Pierre-Gilles reviewed (changes requested):
Hi ! So I really wanted to merge this and right before merging I ran into an issue, just wanted to talk about this with you 😅

When testing the integration, I was navigating on each page of your integration, and I noticed in the logs a:

```
(node:27209) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 device-online listeners added to [Client]. Use emitter.setMaxListeners() to increase limit
```

After investigating it's because of the scan function that add an `.on('device-online'` listener on the client EventEmitter every time the getDevices is called

The thing is that, this listener is added to the global client object, and is never removed, so it creates a memory leak and it means that it'll be slower and slower to query this route as the event listener list get be bigger and bigger

I recommend you remove the listener after the scan is complete! It's a quick fix. See: https://nodejs.org/api/events.html#events_emitter_removelistener_eventname_listener

Sorry I know it's frustrating to find last minute bug but I think it's an important one and we don't want that in production ^^

Thanks again for your time and this really nice PR 🙏

cicoub13 committed:
fix(tp-link): Fix memory leak on eventListeners

Pierre-Gilles reviewed (approved):
Thanks for your changes! No worries, I would have done the same mistake 😄

Congrats again for this PR, i'm good to merge!
